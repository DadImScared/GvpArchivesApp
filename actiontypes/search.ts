export const UPDATE_SEARCH_QUERY = "search/UPDATE_SEARCH_QUERY";
export const TOGGLE_CATEGORIES = "search/TOGGLE_CATEGORIES";
export const SELECT_ALL_CATEGORIES = "search/SELECT_ALL_CATEGORIES";
export const REMOVE_ALL_CATEGORIES = "search/REMOVE_ALL_CATEGORIES";
