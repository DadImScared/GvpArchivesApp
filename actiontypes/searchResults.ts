export const ADD_SEARCH = "searchResults/ADD_SEARCH";
export const UPDATE_SEARCH = "searchResults/UPDATE_SEARCH";
