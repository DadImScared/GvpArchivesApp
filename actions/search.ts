import axios, { AxiosPromise } from "axios";
import { AnyAction, Dispatch } from "redux";
import { action } from "typesafe-actions";

import * as ActionTypes from "../actiontypes/search";
import { AUTOCOMPLETE_QUERY } from "../loadingPrefix";
import { IReducerState } from "../reducers";
import { getQueryId } from "../reducers/search";
import autoComplete from "./autoComplete";
import { BASE_URL } from "./itemsById";

const autoCompleteUrl = `${BASE_URL}/api/v1/completeme/`;

export const isValidValidTimeToLive = (time: number) => time ? time >= +new Date() : false;

const updateSearchQuery = (query: string) => action(ActionTypes.UPDATE_SEARCH_QUERY, query);

type ApiCallCallback<ReturnType> = (
  state: IReducerState,
  dispatch: Dispatch
) => ReturnType;

type IAfterApiCall<Data> = (
  data: Data,
  state: IReducerState,
  dispatch: Dispatch
) => Promise<AnyAction[]> | AnyAction[];

export interface IApiCallConfig<Data> {
  afterApiCall?: IAfterApiCall<Data>;
  apiCall: ApiCallCallback<AxiosPromise<any>>;
  beforeApiCall?: ApiCallCallback<Promise<AnyAction[]> | AnyAction[]>;
  loadingId?: ((state: IReducerState) => string) | string;
  shouldCallApi?: ApiCallCallback<boolean | AxiosPromise<void>>;
}

export const getQuerySuggestions = (query: string, categories: string[]): IApiCallConfig<string[]> => {
  const queryId = getQueryId(query, categories);
  const loadingId = `${AUTOCOMPLETE_QUERY}${queryId}`;
  return {
    loadingId,
    afterApiCall: (data) => {
      return [
        autoComplete.addAutoComplete(
          query,
          categories,
          data.map((textObj: any) => textObj.text)
        )
      ];
    },
    apiCall: () => axios.get(`${autoCompleteUrl}${query}`),
    shouldCallApi: (state) => {
      const isLoading = state.loading[loadingId];
      if (!query || isLoading) return false;

      const currentAutoComplete = state.autoComplete[queryId];
      if (!currentAutoComplete) return true;

      return isValidValidTimeToLive(currentAutoComplete.timeToLive) ?
        false
        :
        axios.get(`${autoCompleteUrl}${query}`);
    }
  };
};

export default {
  updateSearchQuery
};
