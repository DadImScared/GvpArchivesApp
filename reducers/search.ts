import _ from "lodash";
import { ActionType } from "typesafe-actions";

import { search } from "../actions";
import * as ActionTypes from "../actiontypes/search";
import { IReducerState } from "./index";

export interface ISearchState {
  categories: string[];
  query: string;
}

export type SearchStateAndActions<
  T extends keyof ISearchState | keyof typeof search
  > = Pick<ISearchState & typeof search, T>;

export const CATEGORIES = [
  "bhagavatpatrika",
  "book",
  "harikatha",
  "harmonistmagazine",
  "harmonistmonthly",
  "lecture",
  "movie",
  "song"
];

export const getInitialSearchState = (): ISearchState => ({
  categories: CATEGORIES,
  query: ""
});

export default function(state = getInitialSearchState(), action: ActionType<typeof search>) {
  switch (action.type) {
    case ActionTypes.UPDATE_SEARCH_QUERY:
      return {
        ...state,
        query: action.payload
      };
    case ActionTypes.TOGGLE_CATEGORIES:
      const { payload } = action;
      const { categories } = state;
      return {
        ...state,
        categories: categories.includes(payload) ?
          categories.filter((category) => category !== action.payload)
          :
          [...categories, payload]
      };
    case ActionTypes.SELECT_ALL_CATEGORIES:
      return {
        ...state,
        categories: [...CATEGORIES]
      };
    case ActionTypes.REMOVE_ALL_CATEGORIES:
      return {
        ...state,
        categories: []
      };
    default:
      return state;
  }
}

export const getQueryId = (query: string, categories: string[]) => {
  return `${query}-${_.sortBy(categories, (o) => o).join(",")}`;
};

export const getSearch = (state: IReducerState) => state.search;
