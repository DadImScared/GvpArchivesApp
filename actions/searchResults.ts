import axios, { AxiosResponse } from "axios";
import { action } from "typesafe-actions";

import * as ActionTypes from "../actiontypes/searchResults";
import { SEARCH_RESULTS } from "../loadingPrefix";
import { getCategoriesWithDefault } from "../reducers/autoComplete";
import { Item } from "../reducers/itemsById";
import { getQueryId } from "../reducers/search";
import { makeQueryString } from "../utils/queryString";
import { addItemsById, BASE_URL } from "./itemsById";
import { IApiCallConfig, isValidValidTimeToLive } from "./search";

const SEARCH_URL = `${BASE_URL}/api/v1/search`;

export interface ISearchItem extends Item {
  highlightedTitle?: string;
  highlightedBody?: string;
}

interface ISearchResults {
  nextPage: string | boolean;
  results: ISearchItem[];
  suggestions: Array<{ text: string }>;
}

export interface INormalizedSearchResults { item_id: string; highlightedBody?: string; highlightedTitle?: string; }

const addSearch = (
  queryId: string,
  results: INormalizedSearchResults[],
  suggestions: string[],
  nextPage: boolean | string
) => action(ActionTypes.ADD_SEARCH, { queryId, results, suggestions, nextPage });

const updateSearch = (
  queryId: string,
  results: INormalizedSearchResults[],
  suggestions: string[],
  nextPage: boolean | string
) => action(ActionTypes.UPDATE_SEARCH, { queryId, results, suggestions, nextPage });

export const normalizeSearchResults = (items: ISearchItem[]) => {
  const itemsById: { [key: string]: Item } = {};
  const normalizedValues: INormalizedSearchResults[] = [];
  items.forEach(({highlightedBody, highlightedTitle, ...item }) => {
    normalizedValues.push({
      item_id: item.item_id,
      [highlightedBody ? "highlightedBody" : "highlightedTitle"]: highlightedBody || highlightedTitle
    });
    itemsById[item.item_id] = item;
  });
  return {
    itemsById,
    results: normalizedValues
  };
};

export const getSearchResults = (
  query: string,
  selectedCategories: string[],
  refresh = false
): IApiCallConfig<AxiosResponse<ISearchResults>> => {
  const categories = getCategoriesWithDefault(selectedCategories);
  const queryId = getQueryId(query, categories);
  const loadingId = `${SEARCH_RESULTS}${queryId}`;
  const firstPage = `${SEARCH_URL}/${query}/?${makeQueryString({ categories })}`;
  let shouldUpdateSearch = false;
  return {
    loadingId,
    afterApiCall: ({ data }) => {
      // const { result: itemIds, entities: { items } } = normalize(data.results, itemSchema);
      const { itemsById, results } = normalizeSearchResults(data.results);
      return [
        addItemsById(itemsById),
        // add or update search based on bool 'shouldUpdateSearch'
        (shouldUpdateSearch ? updateSearch : addSearch)(
          queryId,
          results,
          data.suggestions.map((suggestion) => suggestion.text),
          data.nextPage
        )
      ];
    },
    apiCall: (state) => axios.get(firstPage),
    shouldCallApi: (state) => {
      if (refresh) return true;
      const isLoading = state.loading[loadingId];

      if (!query || isLoading) return false;

      const currentSearch = state.searchResults[queryId];
      if (!currentSearch) return true;

      if (isValidValidTimeToLive(currentSearch.timeToLive)) {
        if (currentSearch.nextPage) {
          shouldUpdateSearch = true;
          return axios.get(`${BASE_URL}${currentSearch.nextPage}`);
        }
        return false;
      }

      return true;
    },
    type: "callApiMiddleware"
  };
};

export default {
  addSearch,
  updateSearch
};
