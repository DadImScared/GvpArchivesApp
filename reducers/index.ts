import _ from "lodash";
import { combineReducers, DeepPartial } from "redux";

import audioPlayer, { getInitialAudioPlayerState, IAudioPlayerState } from "./audioPlayer";
import autoComplete, { getInitialAutoCompleteState, IAutoCompleteState } from "./autoComplete";
import itemsByCategory, { ItemsByCategoryState } from "./itemsByCategory";
import itemsById, { ItemsByIdState } from "./itemsById";
import loading, { ILoadingState } from "./loading";
import recentSearches, { RecentSearches } from "./recentSearches";
import search, { getInitialSearchState, ISearchState } from "./search";
import searchResults, { ISearchResults } from "./searchResults";

export interface IReducerState {
  audioPlayer: IAudioPlayerState;
  autoComplete: IAutoCompleteState;
  loading: ILoadingState;
  itemsByCategory: ItemsByCategoryState;
  itemsById: ItemsByIdState;
  recentSearches: RecentSearches;
  search: ISearchState;
  searchResults: ISearchResults;
}

export const getInitialReducerState = (injectedState: DeepPartial<IReducerState> = {}) => _.merge({
  audioPlayer: getInitialAudioPlayerState(),
  autoComplete: getInitialAutoCompleteState(),
  itemsByCategory: {},
  itemsById: {},
  loading: {},
  recentSearches: [],
  search: getInitialSearchState(),
  searchResults: {}
}, injectedState);

const reducer = combineReducers<IReducerState>({
  audioPlayer,
  autoComplete,
  itemsByCategory,
  itemsById,
  loading,
  recentSearches,
  search,
  searchResults
});

export default reducer;
