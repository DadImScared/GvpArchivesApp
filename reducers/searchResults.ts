import moment from "moment";
import { ActionType } from "typesafe-actions";

import { NavigationScreenProps } from "react-navigation";
import { createSelector } from "reselect";
import { searchResults } from "../actions";
import { INormalizedSearchResults, ISearchItem } from "../actions/searchResults";
import * as ActionTypes from "../actiontypes/searchResults";
import { SEARCH_RESULTS } from "../loadingPrefix";
import { IReducerState } from "./index";
import { ItemsByIdState } from "./itemsById";

export interface ISearchResult {
  results: INormalizedSearchResults[];
  timeToLive: number;
  suggestions: string[];
  nextPage: string | boolean;
}

export interface ISearchResults {
  [key: string]: ISearchResult;
}

export default function(state: ISearchResults = {}, action: ActionType<typeof searchResults>) {
  switch (action.type) {
    case ActionTypes.ADD_SEARCH:
      const { payload: { queryId, ...other } } = action;
      return {
        ...state,
        [queryId]: {
          ...other,
          timeToLive: moment().add(1, "days").valueOf()
        }
      };
    case ActionTypes.UPDATE_SEARCH:
      const { payload } = action;
      return {
        ...state,
        [payload.queryId]: {
          ...state[payload.queryId],
          nextPage: payload.nextPage,
          results: [...state[payload.queryId].results, ...payload.results]
        }
      };
    default:
      return state;
  }
}

const defaultSearchResult: ISearchResult = {
  nextPage: false,
  results: [],
  suggestions: [],
  timeToLive: 0,
};

export interface ISearchResultData {
  results: ISearchItem[];
  suggestions: string[];
  timeToLive: number;
  isLoading: boolean;
}

export const selectSearchResults = createSelector<
  IReducerState,
  NavigationScreenProps,
  ISearchResult,
  boolean,
  ItemsByIdState,
  ISearchResultData
  >(
  [
    (state, props) => state.searchResults[props.navigation.getParam("queryId")] || defaultSearchResult,
    (state, props) => state.loading[`${SEARCH_RESULTS}${props.navigation.getParam("queryId")}`],
    (state) => state.itemsById
  ],
  (searchResult, isLoading, itemsById) => {
    return {
      isLoading: isLoading === undefined ? true : isLoading,
      results: searchResult.results.map((item) => ({
        ...itemsById[item.item_id],
        ...item
      })),
      suggestions: searchResult.suggestions,
      timeToLive: searchResult.timeToLive,
    };
  }
);

export const makeSelectSearchResults = () => createSelector<
  IReducerState,
  NavigationScreenProps,
  ISearchResult,
  boolean,
  ItemsByIdState,
  ISearchResultData
  >(
  [
    (state, props) => state.searchResults[props.navigation.getParam("queryId")] || defaultSearchResult,
    (state, props) => state.loading[`${SEARCH_RESULTS}${props.navigation.getParam("queryId")}`],
    (state) => state.itemsById
  ],
  (searchResult, isLoading, itemsById) => {
    return {
      isLoading: isLoading === undefined ? true : isLoading,
      results: searchResult.results.map((item) => ({
        ...itemsById[item.item_id],
        ...item
      })),
      suggestions: searchResult.suggestions,
      timeToLive: searchResult.timeToLive,
    };
  }
);
