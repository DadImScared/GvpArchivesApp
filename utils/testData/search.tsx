import * as React from "react";
import * as renderer from "react-test-renderer";

import { action } from "@storybook/addon-actions";
import _ from "lodash";

import { getQueryId } from "../../reducers/search";

export const makeTree = (Component: any, props: any = {}) => (injectedProps: any = {}) => {
  return renderer.create(<Component {...props} {...injectedProps} />);
};

const params = {
  categories: ["book", "song"],
  query: "radha",
  queryId: getQueryId("radha", ["book", "song"])
} as any;

export const getMockFunction = ({
  storybook = false,
  actionName = "",
  cb = false
}: any) => {
  if (storybook) {
    return (...args: any) => {
      action(actionName)(...args);
      if (cb) {
        return cb(...args);
      }
    };
  }
  return cb ? jest.fn(cb) : jest.fn();
};

export const searchTestProps = ({ props = {}, storybook = false } = {}) => (_.merge({
  addRecentSearch: getMockFunction({ storybook, actionName: "addRecentSearch" }),
  getSearchResults: getMockFunction({ storybook, actionName: "getSearchResults" }),
  isLoading: false,
  navigation: {
    getParam: getMockFunction({ storybook, actionName: "getParam", cb: (param: string) => params[param] }),
    push: getMockFunction({ storybook, actionName: "push" })
  },
  results: [
    {
      category: "book",
      item_id: "item-1",
      link: "item-1-link",
      title: "item-1-title"
    }
  ],
  suggestions: ["Suggestion-1", "Suggestion-2", "Suggestion-3"],
  timeToLive: 12345
}, props));
