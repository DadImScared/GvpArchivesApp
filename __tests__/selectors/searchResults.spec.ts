import { getInitialReducerState } from "../../reducers";
import { getQueryId } from "../../reducers/search";
import { makeSelectSearchResults } from "../../reducers/searchResults";

describe("searchResults", () => {
  it("should select search results", () => {
    const queryId = getQueryId("radha", ["book", "movie"]);
    const state = getInitialReducerState(
      {
        itemsById: {
          2: {
            category: "book",
            item_id: "2",
            link: "item2-link",
            title: "item2"
          }
        },
        searchResults: {
          [queryId]: {
            nextPage: false,
            results: [{ item_id: "2", highlightedBody: "search body" }],
            suggestions: [],
            timeToLive: 2
          }
        }
      }
    );
    const props = {
      navigation: {
        getParam: () => queryId
      }
    } as any;
    // console.log(selectSearchResults(state, props));
    expect(makeSelectSearchResults()(state, props)).toMatchSnapshot();
  });
});
