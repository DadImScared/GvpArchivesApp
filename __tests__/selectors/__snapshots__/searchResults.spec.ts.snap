// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searchResults should select search results 1`] = `
Object {
  "isLoading": true,
  "results": Array [
    Object {
      "category": "book",
      "highlightedBody": "search body",
      "item_id": "2",
      "link": "item2-link",
      "title": "item2",
    },
  ],
  "suggestions": Array [],
  "timeToLive": 2,
}
`;
